pyro/contrib/glmm/guides.py

```diff
@@ -69,6 +78,102 @@ def forward(self, y_dict, design, observation_labels, target_labels):
             pyro.sample(l, w_dist)
 
 
+class LinearModelLaplaceGuide(nn.Module):
+    """
+    Laplace approximation for a (G)LM
+    """
+    def __init__(self, d, w_sizes, **kwargs):
+        super(LinearModelLaplaceGuide, self).__init__()
+        # start in train mode
+        self.train()
+        if not isinstance(d, (tuple, list, torch.Tensor)):
+            d = (d,)
+        self.means = {}
+        for l, mu_l in tensor_to_dict(w_sizes, 0.01*torch.ones(*d, sum(w_sizes.values()))).items():
+            self.means[l] = nn.Parameter(mu_l)
+        self._registered = nn.ParameterList(self.means.values())
+        self.scale_trils = {}
+        self.w_sizes = w_sizes
+
+    @staticmethod
+    def _hessian_diag(y, x, event_shape):
+        batch_shape = x.shape[:-len(event_shape)]
+        assert tuple(x.shape) == tuple(batch_shape) + tuple(event_shape)
+
+        dy = torch.autograd.grad(y, [x,], create_graph=True)[0]
+        H = []
+
+        # collapse independent dimensions into a single one,
+        # and dependent dimensions into another single one
+        batch_size = 1
+        for batch_shape_dim in batch_shape:
+            batch_size *= batch_shape_dim
+
+        event_size = 1
+        for event_shape_dim in event_shape:
+            event_size *= event_shape_dim
+
+        flat_dy = dy.reshape(batch_size, event_size)
+
+        # loop over dependent part
+        for i in range(flat_dy.shape[-1]):
+            dyi = flat_dy.index_select(-1, torch.tensor([i]))
+            Hi = torch.autograd.grad([dyi], [x,], grad_outputs=[torch.ones_like(dyi)], retain_graph=True)[0]  # XXX wrong shape
+            H.append(Hi)
+        H = torch.stack(H, -1).reshape(*(x.shape + event_shape))
+        return H
+
+    @staticmethod
+    def _batch_potrf_compat(x, **kwargs):
+        # TODO update this code to Pyro dev and remove in favor of .cholesky()
+        flat_x = x.reshape(-1, x.shape[-2], x.shape[-1])
+        potrfs = torch.stack([flat_xi.potrf(**kwargs) for flat_xi in flat_x])
+        return potrfs.reshape(*tuple(x.shape))
+
+    def finalize(self, loss, target_labels):
+        """
+        Compute the Hessian of the parameters wrt ``loss``
+
+        Usage::
+
+            >>> guide = LinearModelLaplace(...)
+            >>> guide.finalize(loss)
+        """
+        # set self.training = False
+        self.eval()
+        for l, mu_l in self.means.items():
+            if l not in target_labels:
+                continue  # XXX is this right?
```

Comment:
Yes it's right- we shouldn't reach this statement for Laplace though because `target_labels` is meant to match up with the latent variables exactly